### PR TITLE
Colorize checkboxes depending on theming color

### LINKED ISSUE #378: Theming: coloring of folders and log in background
vivaldi_2016-07-12_10-11-08
vivaldi_2016-07-12_10-11-0

Also I think, that **on login page we should use CSS color filter** to color the picture on background. Of course, this feature should work only with default background.
I think it's easy to realise, just convert RGB hex code to HSV and then use three CSS tricks:

```
hue-rotate(0deg);
brightness(0%);
contrast(0%);
```

### PR DESCRIPTION
**Pull request by juliusknorr**
partly fixes #378
- checkboxes match theme color
- checkmark-white.svg replaced to match checkmark.svg

2016-07-15-144933_535x208_scrot
- [x] invert icon color like in #412
- [x] fix unit tests

### CODE REVIEW

**Comment by nickvergessen** [comment]
#412 is merged, want to continue? :grinning:

**Comment by juliushaertl** [comment]
@nickvergessen Sure, I will have some time tomorrow to go on with this.

**Comment by jancborchardt** [comment]
We need to take into account what happens when someone uses white as the color, or very light colors in general which are not very easy to see on the white background. In that case I’d say we either use the default blue or a standard grey.

**Comment by juliushaertl** [comment]
Should be ready for review now.

- Checkbox icon will always stay white, because this looks way better than black, even on bright colors
- Color of the checkboxes will default to grey, if color gets to bright

@nextcloud/designers please check if the boundary for changing to grey is good or if it needs to be changed

I also changed the way the css is built, I hope the string concatenation is not that bad, as the response is cached anyway.

**Comment by nickvergessen** [comment]
Some comments:
* You can hardcode PHP_EOL to `\n`
* If your theme is gray (e.g. cdcdcd) disabled and active boxes look very similar, should we special case this?
* Radio buttons seem to not be covered? Maybe https://github.com/owncloud/core/pull/23063#issuecomment-208293839 helps to cover everything

**Comment by juliushaertl** [comment]
Good point, I will add the other states and radio buttons tomorrow.

At the moment, I don't have a good solution for separating the disabled icon from the checked one, so that they don't look similar with a gray theme color, but I will think about that. Maybe @jancborchardt ?

**Comment by MorrisJobke** [comment]
> At the moment, I don't have a good solution for separating the disabled icon from the checked one, so that they don't look similar with a gray theme color, but I will think about that. Maybe @jancborchardt ?

Then leave it for now as it is. We could iterate on this later.

**Comment by juliushaertl** [comment]
Ok, I will have time after the weekend to continue with this pr.

**Comment by juliushaertl** [comment]
Radio buttons are now also colorized. Please review again.

Normal look:
boxes_1

Look on bright colors:
boxes_2

I'm not sure if generating the SVG for the radio buttons is the best way, at the moment this is more a hacky approach, but my first plan to use mask-image CSS failed, as this is also unsupported in Firefox. @jancborchardt sorry for giving wrong information on that.

**Comment by Henni** [comment]
@juliushaertl have you also experimented with css filters? They should be supported a bit better then masking and you can do some hue shifting magic with it.
[SVG Parameters](https://www.w3.org/TR/SVGParamPrimer/#PassingValuesViaURLParameters) would probably be the best solution. But it is still a working draft, so I doubt that it's widely supported if at all.

What happened to `checkmark-white.svg` and `radio-checked-white.svg`?
Why did you change checkmark-white and why did you add the radio svg?

**Comment by jancborchardt** [comment]
@juliushaertl nice, ok! So currently you basically have an SVG which is the inverse (the white parts) of the icon, and you use background-color to color it, right?

And what do you think about @Henni’s suggestions?

One thing – the color seems different from the actual header color, it’s a bit brighter:
capture du 2016-07-26 17-19-31

File: apps/theming/js/settings-admin.js [hunk 1]
```diff
@@ -46,29 +46,47 @@ function calculateLuminance(rgb) {
...
 function preview(setting, value) {
 	if (setting === 'color') {
 		var headerClass = document.getElementById('header');
 		var expandDisplayNameClass = document.getElementById('expandDisplayName');
 		var headerAppName = headerClass.getElementsByClassName('header-appname')[0];
-		var textColor, icon;
+		var textColor, elementColor, icon;
+
+		var luminance = calculateLuminance(value);
+		var elementColor = value;
```
Comment (MorrisJobke) [inline apps/theming/js/settings-admin.js]: elementColor is already defined 3 lines above.

**Comment by MorrisJobke** [comment]
This has also conflicts. @juliushaertl could I ask you to rebase on latest master and resolve the conflicts? That would be really nice :)

**Comment by juliushaertl** [comment]
@Henni  @jancborchardt 
I've played a bit with css filters, but I could not get them to colorize the elements with the exact theming color. 

> @juliushaertl nice, ok! So currently you basically have an SVG which is the inverse (the white parts) of the icon, and you use background-color to color it, right?

That is the case for the checkbox, the radio button gets generated dynamically. (See https://github.com/nextcloud/server/blob/cc457cd6650c1fd3c1cc6f85c961f65de38615d8/apps/theming/lib/util.php#L78-L82)

> One thing – the color seems different from the actual header color, it’s a bit brighter

This is also the case without the theming app, as these form elements have an
opacity of 0.7 set in core/css/inputs.css. I removed that in the latest commit.

> What happened to checkmark-white.svg and radio-checked-white.svg?
> Why did you change checkmark-white and why did you add the radio svg?

checkmark-white.svg was changed to match the exact form of checkmark.svg
radio-checked-white.svg was added while trying to implement the colorization with mask-image, 
but I think it also makes sense to keep this, as it makes them similar to the checkbox normal/white versions.

**Comment by jancborchardt** [comment]
@juliushaertl very nice! :) So this PR is ready for testing and then good to go from your point of view?

**Comment by juliushaertl** [comment]
Yes, we can change the svg-thing later, if someone has a idea to solve this in a more elegant way.

**Comment by MorrisJobke** [comment]
Nice! Tested and works (IE11, Edge, Firefox, Safari, Chrome)

**Comment by jancborchardt** [comment]
Awesome, works really well! :+1:  :)

**Comment by schiessle** [comment]
@juliushaertl do you want to prepare a backport? Would be nice to have it in Nextcloud 10 and we also need it for https://github.com/nextcloud/server/pull/620 to ship the new first run wizard.

**Comment by juliushaertl** [comment]
@schiessle Yes, but i'll wait until #622 is merged.

**Comment by MorrisJobke** [comment]
> @schiessle Yes, but i'll wait until #622 is merged.

Was merged. @juliushaertl Could you create the backport?

**Pull request merged**